packages/nextjs/utils/abi.ts

```diff
@@ -1,41 +1,88 @@
+    324: process.env.NEXT_PUBLIC_ZKSYNC_ETHERSCAN_API_KEY,
+    137: process.env.NEXT_PUBLIC_POLYGON_ETHERSCAN_API_KEY,
+    42161: process.env.NEXT_PUBLIC_ARBITRUM_ETHERSCAN_API_KEY,
+    534352: process.env.NEXT_PUBLIC_SCROLL_ETHERSCAN_API_KEY,
+    56: process.env.NEXT_PUBLIC_BSC_ETHERSCAN_API_KEY,
+  };
 
-  const url = `https://anyabi.xyz/api/get-abi/${chainId}/${verifiedContractAddress}`;
+  const apiKey = apiKeys[chainId];
 
-  const response = await fetch(url);
-  const data = await response.json();
-  if (data.abi) {
-    return data.abi;
-  } else {
-    console.error("Could not fetch ABI from AnyABI:", data.error);
-    return;
+  if (!apiKey) {
+    console.warn(`No API key found for chain ID ${chainId}`);
   }
+
+  return apiKey || "";
 };
 
-export const fetchContractABIFromEtherscan = async (verifiedContractAddress: string, chainId: number) => {
-  const chain = NETWORKS_EXTRA_DATA[chainId];
+export const fetchContractABIFromEtherscan = async (verifiedContractAddress: Address, chainId: number) => {
+  const chain = findChainById(chainId);
 
-  if (!chain || !chain.etherscanEndpoint)
-    throw new Error(`ChainId ${chainId} not found in supported etherscan networks`);
+  if (!chain || !chain.blockExplorers?.default?.apiUrl) {
+    throw new Error(`ChainId ${chainId} not found in supported networks or missing block explorer API URL`);
+  }
 
-  const apiKey = chain.etherscanApiKey ?? "";
+  const apiKey = getEtherscanApiKey(chainId);
   const apiKeyUrlParam = apiKey.trim().length > 0 ? `&apikey=${apiKey}` : "";
-  const url = `${chain.etherscanEndpoint}/api?module=contract&action=getabi&address=${verifiedContractAddress}${apiKeyUrlParam}`;
-
-  const response = await fetch(url);
-  const data = await response.json();
-  if (data.status === "1") {
-    return data.result;
-  } else {
-    console.error("Got non-1 status from Etherscan API", data);
-    if (data.result) throw new Error(data.result);
-    throw new Error("Got non-1 status from Etherscan API");
+
+  // First call to get source code and check for implementation
+  const sourceCodeUrl = `${chain.blockExplorers.default.apiUrl}?module=contract&action=getsourcecode&address=${verifiedContractAddress}${apiKeyUrlParam}`;
+
+  console.log("Source code URL:", sourceCodeUrl);
+
+  const sourceCodeResponse = await fetch(sourceCodeUrl);
+  const sourceCodeData = await sourceCodeResponse.json();
+
+  if (sourceCodeData.status !== "1" || !sourceCodeData.result || sourceCodeData.result.length === 0) {
+    console.error("Error fetching source code from Etherscan:", sourceCodeData);
+    throw new Error("Failed to fetch source code from Etherscan");
   }
+
+  const contractData = sourceCodeData.result[0];
+  const implementation = contractData.Implementation || null;
+
+  // If there's an implementation address, make a second call to get its ABI
+  if (implementation && implementation !== "0x0000000000000000000000000000000000000000") {
```

Comment:
NIT: I think maybe we could have https://github.com/scaffold-eth/scaffold-eth-2/blob/e6b84bf3e0a2eba7ae3d0ed47d66198aba1a3968/packages/nextjs/utils/scaffold-eth/common.ts#L7 from SE-2 since we are checking it at different places